### ISSUE TITLE
GetUserAttributeVerficationCode and VerifyUser method implementation. [proposed label] feature-request

### ISSUE THREAD

**Issue by vishalkhedeker** [aws-amplify/amplify-flutter#430]
Hi, i couldnt find getUserAttibuteVerificationCode and verifyAttribute methods in amplify_auth_cognito for verfiying more than one attribute.
Will this be added to amplify-flutter?
If so by when is it expected?

**Comment by haverchuck** [comment]
@vishalkhedeker I don't have an exact date, but this is definitely on our roadmap!

**Comment by hysuki** [comment]
@vishalkhedeker Until it is officially supported, it can be implemented through api and function with accessToken if necessary.

For example
**API**

```graphql
type Mutation {
    verifyPhoneNumber(access_token: String, username: String, phone_number: String): verifyPhoneNumberResponse @function(name: "verifyPhoneNumber-${env}")
    verifyPhoneNumberConfirm(access_token: String, code: String): verifyPhoneNumberResponse @function(name: "verifyPhoneNumberConfirm-${env}")
}
```

```dart
class VerifyPhoneNumber {
  var logger = Logger();

  String graphQLDocument = '''
  mutation VerifyPhoneNumber(\$access_token: String = "", \$phone_number: String = "", \$username: String = "") {
    verifyPhoneNumber(access_token: \$access_token, phone_number: \$phone_number, username: \$username) {
      status
    }
  }
  ''';

  final String accessToken;
  final String phoneNumber;
  final String username;

  VerifyPhoneNumber({
    @required this.accessToken,
    @required this.phoneNumber,
    @required this.username,
  })  : assert(accessToken != null),
        assert(phoneNumber != null),
        assert(username != null);

  Future<String> data() async {
    try {
      var operation = Amplify.API.mutate(request: GraphQLRequest<String>(document: graphQLDocument, variables: {
        'access_token': accessToken,
        'phone_number': phoneNumber,
        'username': username,
      }));

      var response = await operation.response;
      var data = response.data;

      logger.d(data);
      return data;
    } on ApiException catch (e) {
      logger.e(e.message);
      return Future.error(e.message);
    }
  }
}




class VerifyPhoneNumberConfirm {
  var logger = Logger();

  String graphQLDocument = '''
  mutation VerifyPhoneNumberConfirm(\$access_token: String = "", \$code: String = "") {
    verifyPhoneNumberConfirm(access_token: \$access_token, code: \$code) {
      status
    }
  }
  ''';

  final String accessToken;
  final String code;

  VerifyPhoneNumberConfirm({
    @required this.accessToken,
    @required this.code,
  })  : assert(accessToken != null),
        assert(code != null);

  Future<String> data() async {
    try {
      var operation = Amplify.API.mutate(request: GraphQLRequest<String>(document: graphQLDocument, variables: {
        'access_token': accessToken,
        'code': code,
      }));

      var response = await operation.response;
      var data = response.data;

      logger.d(data);
      return data;
    } on ApiException catch (e) {
      logger.e(e.message);
      return Future.error(e.message);
    }
  }
}

.............


// Fetch Session
  Future<CognitoAuthSession> fetchSession() async {
    try {
      CognitoAuthSession session = await Amplify.Auth.fetchAuthSession(
          options: CognitoSessionOptions(getAWSCredentials: true));
      logger.d(session);
      return session;
    } on AuthException catch (e) {
      logger.e(e.message);
      return Future.error(e.message);
    }
  }

// Get AccessToken
  Future<String> get accessToken async {
    CognitoAuthSession session = await fetchSession();
    return session.userPoolTokens.accessToken;
  }

// Send verify SMS
Future<bool> verifyPhoneNumber(String phoneNumber) async {
    try {
      Mutations.verifyPhoneNumber = VerifyPhoneNumber(accessToken: await accessToken, phoneNumber: phoneNumber, username:await username);
      String data = await Mutations.verifyPhoneNumber.data();
      return true;
    } catch (e) {
      logger.d(e);
      return false;
    }
  }

// Confirm code
  Future<bool> verifyPhoneNumberConfirm(String code) async {
    try {
      Mutations.verifyPhoneNumberConfirm = VerifyPhoneNumberConfirm(accessToken: await accessToken, code: code);
      String data = await Mutations.verifyPhoneNumberConfirm.data();
      return true;
    } catch (e) {
      logger.d(e);
      return false;
    }
  }
```

In function

**For verify PhoneNumber**

```go
package main

import (
	"context"
	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider/types"
	"log"

	"github.com/aws/aws-lambda-go/lambda"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
)

const userPoolID = "us-east-1_xxxxxxxxxxxxx"

type request struct {
	Arguments arguments `json:"arguments"`
}

type arguments struct {
	AccessToken string `json:"access_token"`
	Username string `json:"username"`
	PhoneNumber string `json:"phone_number"`
}

type response struct {
	Status bool `json:"status"`
}

func HandleRequest(ctx context.Context, req request) (*response, error) {
	resp, err := verifyPhoneNumberRequest(req.Arguments)
	if err != nil {
		log.Fatal(err)
		return nil, err
	}
	return resp, nil
}

func main() {
	lambda.Start(HandleRequest)
}

func verifyPhoneNumberRequest(args arguments) (*response, error) {
	cfg, err := config.LoadDefaultConfig(context.TODO())
	if err != nil {
		return nil, err
	}

	client := cognitoidentityprovider.NewFromConfig(cfg)

	_, err = client.AdminUpdateUserAttributes(context.Background(), &cognitoidentityprovider.AdminUpdateUserAttributesInput{
		UserAttributes: []types.AttributeType{
			{
				Name:  aws.String("phone_number"),
				Value: aws.String(args.PhoneNumber),
			},
		},
		UserPoolId: aws.String(userPoolID),
		Username:   aws.String(args.Username),
	})
	if err != nil {
		return nil, err
	}

	_, err = client.GetUserAttributeVerificationCode(context.Background(), &cognitoidentityprovider.GetUserAttributeVerificationCodeInput{
		AccessToken:    aws.String(args.AccessToken),
		AttributeName:  aws.String("phone_number"),
		ClientMetadata: nil,
	})

	if err != nil {
		return nil, err
	}

	res := &response{Status: true}
	return res, nil
}

```

**For Confirm code**

```go
package main

import (
  "context"
  "github.com/aws/aws-lambda-go/lambda"
  "log"

  "github.com/aws/aws-sdk-go-v2/aws"
  "github.com/aws/aws-sdk-go-v2/config"
  "github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
)

const userPoolID = "us-east-1_xxxxxxxxx"

type request struct {
  Arguments arguments `json:"arguments"`
}

type arguments struct {
  AccessToken string `json:"access_token"`
  Code string `json:"code"`
}

type response struct {
  Status bool `json:"status"`
}

func HandleRequest(ctx context.Context, req request) (*response, error) {
  resp, err := verifyPhoneNumberConfirm(req.Arguments)
  if err != nil {
    log.Fatal(err)
    return nil, err
  }
  return resp, nil
}

func main() {
  lambda.Start(HandleRequest)
}

func verifyPhoneNumberConfirm(args arguments) (*response, error) {
  cfg, err := config.LoadDefaultConfig(context.TODO())
  if err != nil {
    return nil, err
  }

  client := cognitoidentityprovider.NewFromConfig(cfg)
  _, err = client.VerifyUserAttribute(context.Background(), &cognitoidentityprovider.VerifyUserAttributeInput{
    AccessToken:   aws.String(args.AccessToken),
    AttributeName: aws.String("phone_number"),
    Code:          aws.String(args.Code),
  })

  if err != nil {
    return nil, err
  }

  res := &response{Status: true}

  return res, nil
}

```

**Comment by Jordan-Nelson** [comment]
Hi @vishalkhedeker - The APIs to update user attributes are now available in version 0.1.5. The [docs on user attributes](https://docs.amplify.aws/lib/auth/user-attributes/q/platform/flutter#update-user-attribute) have been updated to include the new APIs. I am going to close out this issue as I believe those APIs are what you are looking for.
